code/modules/projectiles/guns/beam_rifle.dm

```diff
@@ -85,149 +180,173 @@
 /obj/item/weapon/gun/energy/beam_rifle/attack_self(mob/user)
 	projectile_setting_pierce = !projectile_setting_pierce
 	to_chat(user, "<span class='boldnotice'>You set \the [src] to [projectile_setting_pierce? "pierce":"impact"] mode.</span>")
+	aiming_beam()
 
 /obj/item/weapon/gun/energy/beam_rifle/proc/update_slowdown()
-	if(scoped)
+	if(aiming)
 		slowdown = scoped_slow
 	else
 		slowdown = initial(slowdown)
 
 /obj/item/weapon/gun/energy/beam_rifle/Initialize()
 	. = ..()
 	START_PROCESSING(SSfastprocess, src)
+	zoom_speed_action = new(src)
+	zoom_lock_action = new(src)
 
 /obj/item/weapon/gun/energy/beam_rifle/Destroy()
 	STOP_PROCESSING(SSfastprocess, src)
+	set_user(null)
+	clear_tracer()
 	..()
 
-/obj/item/weapon/gun/energy/beam_rifle/zoom(user, forced_zoom)
-	. = ..()
-	scope(user, .)
-
-/obj/item/weapon/gun/energy/beam_rifle/proc/scope(mob/user, forced)
-	var/scoping
-	switch(forced)
-		if(TRUE)
-			scoping = TRUE
-		if(FALSE)
-			scoping = FALSE
-		else
-			scoping = !scoped
-	if(scoping)
-		spread = scoped_inaccuracy
-		recoil = scoped_recoil
-		scoped = TRUE
-		user << "<span class='boldnotice'>You bring your [src] up and use its scope...</span>"
-	else
-		spread = hipfire_inaccuracy
-		recoil = hipfire_recoil
-		scoped = FALSE
-		user << "<span class='boldnotice'>You lower your [src].</span>"
-	update_slowdown()
-
-/obj/item/weapon/gun/energy/beam_rifle/can_trigger_gun(var/mob/living/user)
-	if(!scoped && !noscope)
-		user << "<span class='userdanger'>This beam rifle can only be used while scoped!</span>"
-		return FALSE
-	. = ..(user)
-
 /obj/item/weapon/gun/energy/beam_rifle/emp_act(severity)
 	chambered = null
 	recharge_newshot()
 
-/obj/item/weapon/gun/energy/beam_rifle/proc/aiming_beam()
-	var/atom/A = current_user.client.mouseObject
-	if(!istype(A) || !A.loc)
-		return
-	var/turf/T = get_turf(current_user.client.mouseObject)
-	if(!istype(T))
+/obj/item/weapon/gun/energy/beam_rifle/proc/aiming_beam(force_update = FALSE)
+	var/diff = abs(aiming_lastangle - lastangle)
+	check_user()
+	if(diff < AIMING_BEAM_ANGLE_CHANGE_THRESHOLD && !force_update)
 		return
+	aiming_lastangle = lastangle
+	CHECK_TICK
```

Comment:
leftovers from before when i tried making 1500 objects per second and felt it was laggy but it's not needed now i'll remove it